core/src/main/java/io/grpc/internal/CallCredentialsApplyingTransportFactory.java

```diff
@@ -126,5 +140,50 @@ public Attributes getTransportAttrs() {
         return delegate.newStream(method, headers, callOptions);
       }
     }
+
+    @Override
+    public Runnable start(Listener listener) {
+      this.listener = new CallCredentialsApplyingTransportListener(listener);
+      return super.start(this.listener);
+    }
+    // TODO(zivy@): cancel pending appliers when shutdownNow.
+
+    private class CallCredentialsApplyingTransportListener implements Listener {
+      private final Listener delegateListener;
+      private volatile boolean savedTransportTerminated;
+
+      public CallCredentialsApplyingTransportListener(Listener listener) {
+        this.delegateListener = listener;
+      }
+
+      @Override
+      public void transportShutdown(Status s) {
+        delegateListener.transportShutdown(s);
+      }
+
+      @Override
+      public void transportTerminated() {
+        savedTransportTerminated = true;
+        maybeTerminated();
+      }
+
+      @Override
+      public void transportReady() {
+        delegateListener.transportReady();
+      }
+
+      @Override
+      public void transportInUse(boolean inUse) {
+        delegateListener.transportInUse(inUse);
+      }
+
+      public void maybeTerminated() {
+        synchronized (this) {
+          if (pendingApplier.get() == 0 && savedTransportTerminated) {
```

Comment:
This fixes the executor problem but still has a problem that this will continue trying to create streams after it shut down the delegate. I think you need to delay `shutdown()` until `pendingApplier.get() == 0`. 

shutdownNow() is more trouble, as we do want to actually call shutdownNow on the delegate. So you will still need something like this.

Reply:
> This fixes the executor problem but still has a problem that this will continue  trying to create streams after it shut down the delegate.

I guess this PR does actually does not fix the executor problem. That problem will be fixed by #7750, which also make sure no new stream on callCredentailsTransport after shutdown.

Right, it seems I have to delay shutdown() not terminate(), so that no new stream on real transport after shutdown.

Reply:
#7750 is separate. It fixes the same problem, but is unrelated to this code. If you assume it is fixed/not a problem when considering this code, then you'll see there's still a problem here.

The "executor problem" here just involves the metadata applier. And I do think this code addresses the problem for this class. But the problem is making us look deeper at what is going on and there are other lifecycle issues.

Reply:
> The "executor problem" here just involves the metadata applier. And I do think this code addresses the problem for this class. But the problem is making us look deeper at what is going on and there are other lifecycle issues.

It looks it depends how internalSubchannel() and the executor is shutdown in ManagedChannelImpl, as this class may be executed on the executor which out of its own control.

I find the "Cancellation and Shutdown" chapter in the [practice book](https://www.goodreads.com/book/show/127932.Java_Concurrency_in_Practice) pretty relevant, but this class indeed is by no means a complete lifecycle.